### PR TITLE
ControlBoardDriver: add iJointCoupling interface

### PR DESCRIPTION
**Pull request by Nicogene**
This PR adds the [`iJointCoupling` interface](https://github.com/robotology/yarp/blob/master/src/libYARP_dev/src/yarp/dev/IJointCoupling.h), that allows to handle coupled systems using YARP device driver such as:
- https://github.com/icub-tech-iit/ergocub-software/blob/master/src/modules/couplingXCubHandMk5/README.md

### CODE REVIEW

**Comment by Nicogene** [comment]
The latest commit seems a reasonable implementation, it is all base on this line of code (that I hope it works as I think):
```c++
    m_controlBoardData.actuated_joints_handle = m_controlBoardData.ijointcoupling ? &(m_controlBoardData.actuatedAxes) : &(m_controlBoardData.physicalJoints); 
```

Because I introduced a second set of joints (physicalJoints and actuatedAxes) and in function of the presence or not of coupling they are the same thing or not.

For now I didn't introduced the decoupling of the velocities because it seems not used

**Comment by Nicogene** [comment]
Yesterday @traversaro @xela-95 and I had an alignment, and we decided to deeply refactor the plugin in order to make cleaner the introduction of the coupling.

I managed to have a first version that compiles, it is already a success given how many things I changed :D 

I will now test it using the current configuration files of ergocub (no coupling involved), in case of positive outcome I will move to test it with the coupling of the hands mk5.x.

**Comment by Nicogene** [comment]
> If you want a more self-contained test with a smaller model w.r.t. to ergocub, it seems that also the single_pendulum tests is currently failing.

Thanks for the heads up! I used that test for debugging it and I managed to find that specific error, one controlboard test is still failing I am not sure if it is due to my changes

**Comment by Nicogene** [comment]
I managed to fix:
- limits
- names

for the actuated axes, I am still getting a segmentation fault on `ControlBoardCommonsTest`

At first glance, ergocub seems controllable with my changes
image

**Comment by Nicogene** [comment]
The coupling seems to work just fine

image

**Comment by Nicogene** [comment]
I noticed that enabling the control of the hands make drop the RTF of a ~ 20%, this can be noticed also in the screenshots I posted:
- [No hands](https://github.com/robotology/gz-sim-yarp-plugins/pull/221#issuecomment-2541752185) RTF **79%**
- [With hands](https://github.com/robotology/gz-sim-yarp-plugins/pull/221#issuecomment-2545211545) RTF **58%**

I have also checked the RTF on `main` and it is ~ 78% so it seems not that my changes slowed down the simulation, probably the more joints to control make the simulation slower


cc @xela-95 @traversaro

**Comment by Nicogene** [comment]
> Great! Not blocking, but most of the (implicit) documentation for plugins is given in the `tutorial` folder in form of examples for each plugin. To have an example for the coupling functionality, we could port [robotology/gazebo-yarp-plugins@`master`/tutorial/model/coupled_pendulum](https://github.com/robotology/gazebo-yarp-plugins/tree/master/tutorial/model/coupled_pendulum) (and obviously it will start depending on icub-main). That will provide also a simple smoke test for the coupling functionality in
> 
> https://github.com/robotology/gz-sim-yarp-plugins/blob/64deb83de3d4bbb606c14ee6f3605212b1004587/tests/gz-sim-yarp-plugins-check-model/CMakeLists.txt#L19
> .

Ok for me! But we need first to merge the coupling device PR in icub-main:
- https://github.com/robotology/icub-main/pull/997

**Comment by traversaro** [comment]
> > Great! Not blocking, but most of the (implicit) documentation for plugins is given in the `tutorial` folder in form of examples for each plugin. To have an example for the coupling functionality, we could port [robotology/gazebo-yarp-plugins@`master`/tutorial/model/coupled_pendulum](https://github.com/robotology/gazebo-yarp-plugins/tree/master/tutorial/model/coupled_pendulum) (and obviously it will start depending on icub-main). That will provide also a simple smoke test for the coupling functionality in
> > https://github.com/robotology/gz-sim-yarp-plugins/blob/64deb83de3d4bbb606c14ee6f3605212b1004587/tests/gz-sim-yarp-plugins-check-model/CMakeLists.txt#L19
> > 
> > .
> 
> Ok for me! But we need first to merge the coupling device PR in icub-main:
> 
>     * [Add `couplingICubEye` device icub-main#997](https://github.com/robotology/icub-main/pull/997)

Sure! We can also just add tutorial and then add the tests once https://github.com/robotology/icub-main/pull/997 is merged.

**Comment by Nicogene** [comment]
> > > Great! Not blocking, but most of the (implicit) documentation for plugins is given in the `tutorial` folder in form of examples for each plugin. To have an example for the coupling functionality, we could port [robotology/gazebo-yarp-plugins@`master`/tutorial/model/coupled_pendulum](https://github.com/robotology/gazebo-yarp-plugins/tree/master/tutorial/model/coupled_pendulum) (and obviously it will start depending on icub-main). That will provide also a simple smoke test for the coupling functionality in
> > > https://github.com/robotology/gz-sim-yarp-plugins/blob/64deb83de3d4bbb606c14ee6f3605212b1004587/tests/gz-sim-yarp-plugins-check-model/CMakeLists.txt#L19
> > > 
> > > .
> > 
> > 
> > Ok for me! But we need first to merge the coupling device PR in icub-main:
> > ```
> > * [Add `couplingICubEye` device icub-main#997](https://github.com/robotology/icub-main/pull/997)
> > ```
> 
> Sure! We can also just add tutorial and then add the tests once [robotology/icub-main#997](https://github.com/robotology/icub-main/pull/997) is merged.

I would prefer to add the tutorial after merging https://github.com/robotology/icub-main/pull/997 because I am not 100% about the configuration file for that coupling, I can add the tutorial but it is possible that the configuration files has  to be revised after testing the coupling of the eyes

**Comment by traversaro** [comment]
Ok, no problem for me!

File: plugins/controlboard/include/ControlBoardData.hh
Comment (xela-95) [inline plugins/controlboard/include/ControlBoardData.hh:65]: Since this is referred to a single axis maybe we should maintain the singular?

```suggestion
struct ActuatedAxisProperties
```

File: plugins/controlboard/src/ControlBoard.cpp
Comment (traversaro) [inline plugins/controlboard/src/ControlBoard.cpp:137]: Can we avoid to change the `m_pluginParameters` attribute? If we need to get all the parameters of `m_pluginParameters`, we can copy this to a `m_couplingParameters` attribute, and then set the device just there. However, just to understand, why do we need to pass the whole plugin parameters to the coupling device?

File: plugins/controlboard/src/ControlBoard.cpp
Comment (traversaro) [inline plugins/controlboard/src/ControlBoard.cpp:143]:
```suggestion
                        "coupling driver with device name " << coupling_device_str 
```

I think this may be useful for debug if something does not work.

File: plugins/controlboard/src/ControlBoardData.cpp
Comment (traversaro) [inline plugins/controlboard/src/ControlBoardData.cpp:15]: Can we cache this in an attribute buffer instead of calling it every time?

File: plugins/controlboard/src/ControlBoard.cpp
Comment (traversaro) [inline plugins/controlboard/src/ControlBoard.cpp:268]: This function can fail, can we propagate the error to the caller?

**Review by traversaro — CHANGES_REQUESTED** [review]
Minor comments, the one that I care more is to avoid to change the content of `m_pluginParameters`

File: plugins/controlboard/src/ControlBoardData.cpp
Comment (xela-95) [inline plugins/controlboard/src/ControlBoardData.cpp:2]: Just to be consistent with the codebase

```suggestion
#include <ControlBoardData.hh>
```

File: plugins/controlboard/include/ControlBoardData.hh
Comment (xela-95) [inline plugins/controlboard/include/ControlBoardData.hh:82]: Can we try to use a smart pointer like `std::shared_ptr` instead of raw pointers?

File: plugins/controlboard/src/ControlBoard.cpp
Comment (xela-95) [inline plugins/controlboard/src/ControlBoard.cpp:440]: Leftover?
```suggestion
```

File: plugins/controlboard/include/ControlBoardData.hh
Comment (Nicogene) [reply plugins/controlboard/include/ControlBoardData.hh:83]: mmm I am not a great fan of using smart pointers w/ YARP interfaces, because the memory ownership when you do the `view` is not very clear to me

**Review by xela-95 — APPROVED** [review]
Thanks a lot @Nicogene for the great work! It looks good to me! 🚀 

Maybe it could be useful to add the coupled pendulum tutorial as suggested by @traversaro both as an example for the user and as a smoke test for the coupling feature. But this can be done in another PR.

File: plugins/controlboard/include/ControlBoardData.hh
Comment (traversaro) [reply plugins/controlboard/include/ControlBoardData.hh:83]: Yes, I can confirm. If you need to access a pointer that is not managed as smart point, you can't use smart pointers.

File: plugins/controlboard/src/ControlBoard.cpp
Comment (traversaro) [reply plugins/controlboard/src/ControlBoard.cpp:137]: Ack, can you add a comment explaining that and avoid to change the content of `m_pluginParameters` and create a new `yarp::os::Property` instance instead? Thanks!

File: plugins/controlboard/include/ControlBoardData.hh
Comment (xela-95) [reply plugins/controlboard/include/ControlBoardData.hh:82]: Ah ok thanks understood now!

**Comment by Nicogene** [comment]
I should have addressed all your requested changes, and I re-tested the coupling after the changes

**Review by traversaro — APPROVED** [review]
Great, thanks a lot!

**Comment by traversaro** [comment]
@xela-95 the PR for me is ready to merge once CI is happy. I guess we can merge with squash as the history is not curated.

**Pull request merged**
